031-custom_authentication.md

```diff
@@ -0,0 +1,104 @@
+
+## Proposal
+
+Add a CustomAuthenticationKafkaListener class, which supports an example of the following properties
+
+```
+listener.name.<listener-name>.oauthbearer.sasl.client.callback.handler.class=
+listener.name.<listener-name>.oauthbearer.sasl.server.callback.handler.class=
+listener.name.<listener-name>.oauthbearer.sasl.login.callback.handler.class=
+listener.name.<listener-name>.oauthbearer.connections.max.reauth.ms=
+listener.name.<listener-name>.sasl.enabled.mechanisms=
+listener.name.<listener-name>.oauthbearer.sasl.jaas.config=
+listener.name.<listener-name>.AWS_MSK_IAM.sasl.jaas.config=
+listener.name.<listener-name>.AWS_MSK_IAM.sasl.client.callback.handler.class=
+listener.security.protocol.map=
+principal.builder.class=
+```
+
+Really, this can be boiled down to: 
+
+```
+listener.name.<lister-name>.<auth-type>.*
+listener.name.<listener-name>.sasl.enabled.mechanisms
+security.protocol
+principal.builder.class
+```
+
+Taking this further, the desired yaml should look like: 
+
+```
+    listeners:
+      - name: tls
+        port: 9093
+        type: internal
+        tls: true
+        authentication:
+          type: custom
+          principal.builder.class: SimplePrincipal.class
+          security.protocol: SASL_SSL
+          sasl.enabled.mechanisms:
+            - OAUTHBEARER
+          listener-config:
+            oauthbearer.sasl.client.callback.handler.class: client.class
+            oauthbearer.sasl.server.callback.handler.class: server.class
+            oauthbearer.sasl.login.callback.handler.class: login.class
+            oauthbearer.connections.max.reauth.ms: 999999999
+          tlsTrustedCertificates: ...
+```
+
+Then, when constructing the broker config:
+
+* `principal.builder.class` is set cluster-wide for all authentication methods (this is the status quo for OAuth authentication). This can only be set once, and we will fail hard (i.e. reject the config) if there are multiple principals defined. 
+    
+    It is expected that the provided principal builder *must* support Strimzi authentication (i.e. ssl-based auth). This is expected to be identical, or very similar to [Strimzi's OAuth library](https://github.com/strimzi/strimzi-kafka-oauth/blob/main/oauth-server/src/main/java/io/strimzi/kafka/oauth/server/OAuthKafkaPrincipalBuilder.java). In particular, it leverages [Kafka's default principal builder class](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/security/authenticator/DefaultKafkaPrincipalBuilder.java#L73-L79), which allows the building of a principal based upon the name of the of peer certificate. Thus, it is expected that the custom principal builder provides a principal of type "user" with the name being that of the ssl peer certificate. 
+
+    In addition, it also expected that the custom authorizer supports principals of type "user", and loads the `super.users` property to know which user's (i.e. Strimzi services) are authorized. 
```

Comment:
Not sure I understand how does this fit into it this. The authorization is independent on authentication, so this seems a bit random here and unrelated to content around it. Maybe you can clarify it?